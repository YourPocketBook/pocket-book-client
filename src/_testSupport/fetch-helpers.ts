import fetchMock, {
  MockOptionsMethodDelete,
  MockOptionsMethodGet,
  MockOptionsMethodPost,
  MockOptionsMethodPut,
  MockResponse,
  MockResponseFunction
} from "fetch-mock";

afterEach(fetchMock.reset);

export function authenticatedGet(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodGet
) {
  fetchMock.get(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}

export function authenticatedPut(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPut
) {
  fetchMock.put(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedPost(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPost
) {
  fetchMock.post(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedDelete(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodDelete
) {
  fetchMock.delete(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}
