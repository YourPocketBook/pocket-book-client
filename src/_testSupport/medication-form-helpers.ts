import { fireEvent } from "@testing-library/react";

export const testDefaultState = {
  adviceIfDeclined: "Test Advice 1",
  adviceIfTaken: "Test Advice 2",
  dose: "Test Dose",
  exclusionCriteria: "Test Exclusion",
  form: "Test Form",
  inclusionCriteria: "Test Inclusion",
  indications: "Test Indications",
  name: "Test Name",
  policyDate: "2019-12-10",
  route: "Test Route",
  sideEffects: "Test Side Effects"
};

export function setupState(getByLabelText: (text: string) => HTMLElement) {
  fireEvent.change(getByLabelText("Name"), {
    target: { value: testDefaultState.name }
  });
  fireEvent.change(getByLabelText("Policy Date"), {
    target: { value: testDefaultState.policyDate }
  });
  fireEvent.change(getByLabelText("Indications"), {
    target: { value: testDefaultState.indications }
  });
  fireEvent.change(getByLabelText("Inclusion Criteria"), {
    target: { value: testDefaultState.inclusionCriteria }
  });
  fireEvent.change(getByLabelText("Exclusion Criteria"), {
    target: { value: testDefaultState.exclusionCriteria }
  });
  fireEvent.change(getByLabelText("Name, Forms and Strengths of Medicine"), {
    target: { value: testDefaultState.form }
  });
  fireEvent.change(getByLabelText("Route or Method"), {
    target: { value: testDefaultState.route }
  });
  fireEvent.change(getByLabelText("Dosage"), {
    target: { value: testDefaultState.dose }
  });
  fireEvent.change(getByLabelText("Side Effects"), {
    target: { value: testDefaultState.sideEffects }
  });
  fireEvent.change(getByLabelText("Advice if Taken"), {
    target: { value: testDefaultState.adviceIfTaken }
  });
  fireEvent.change(getByLabelText("Advice if Declined"), {
    target: { value: testDefaultState.adviceIfDeclined }
  });
}

export function checkForClass(
  getByLabelText: (text: string) => HTMLElement,
  cls: "is-valid" | "is-invalid"
) {
  expect(getByLabelText("Name")).toHaveClass(cls);
  expect(getByLabelText("Policy Date")).toHaveClass(cls);
  expect(getByLabelText("Indications")).toHaveClass(cls);
  expect(getByLabelText("Inclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Exclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Name, Forms and Strengths of Medicine")).toHaveClass(
    cls
  );
  expect(getByLabelText("Route or Method")).toHaveClass(cls);
  expect(getByLabelText("Dosage")).toHaveClass(cls);
  expect(getByLabelText("Side Effects")).toHaveClass(cls);
  expect(getByLabelText("Advice if Taken")).toHaveClass(cls);
  expect(getByLabelText("Advice if Declined")).toHaveClass(cls);
}

export function checkState(
  getByLabelText: (text: string) => HTMLElement,
  hasValue: boolean
) {
  expect(getByLabelText("Name")).toHaveValue(
    hasValue ? testDefaultState.name : ""
  );
  expect(getByLabelText("Policy Date")).toHaveValue(
    hasValue ? testDefaultState.policyDate : ""
  );
  expect(getByLabelText("Indications")).toHaveTextContent(
    hasValue ? testDefaultState.indications : ""
  );
  expect(getByLabelText("Inclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.inclusionCriteria : ""
  );
  expect(getByLabelText("Exclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.exclusionCriteria : ""
  );
  expect(
    getByLabelText("Name, Forms and Strengths of Medicine")
  ).toHaveTextContent(hasValue ? testDefaultState.form : "");
  expect(getByLabelText("Route or Method")).toHaveTextContent(
    hasValue ? testDefaultState.route : ""
  );
  expect(getByLabelText("Dosage")).toHaveTextContent(
    hasValue ? testDefaultState.dose : ""
  );
  expect(getByLabelText("Side Effects")).toHaveTextContent(
    hasValue ? testDefaultState.sideEffects : ""
  );
  expect(getByLabelText("Advice if Taken")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfTaken : ""
  );
  expect(getByLabelText("Advice if Declined")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfDeclined : ""
  );
}
