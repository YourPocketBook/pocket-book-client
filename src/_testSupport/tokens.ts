export const inDateNonAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.VmT5YkTG0sUd5sK7skfMdi9kc5XG7Yhx6sqhqMSxhVY";
export const inDateAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInJvbGUiOiJBZG1pbiIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.rSQb2Hv6RfxaXSfO6s3QjLjnYKeqVgps3rPenquh3Nk";
export const expiredToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE0MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInVuaXF1ZV9uYW1lIjoiVGVzdCBVc2VyIiwibmFtZWlkIjoiZmYxZDA4ZTctNDY1Ni00YzI4LTk3NmYtMWMyN2YxOTE0NzdmIiwiZW1haWwiOiJ0ZXN0QHNqYS5vcmcudWsifQ.7bDfHDk9sWXL8boFeZWIdo_GB-4Qj8PvA4HDC49EzNc";
