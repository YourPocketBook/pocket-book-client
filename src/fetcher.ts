import isFuture from "date-fns/isFuture";
import jwtDecode from "jwt-decode";

class Fetcher {
  private realName: string = "";
  private nameChanged?: (name: string) => void;
  private token: string | null;
  private parsedToken:
    | { exp: number; given_name: string; role: string }
    | undefined;

  constructor() {
    this.token = localStorage.getItem("token");
    this.parseToken();
  }

  public setNameChangeHandler(func?: (name: string) => void) {
    this.nameChanged = func;
  }

  public authFetch(input: RequestInfo, init?: RequestInit): Promise<Response> {
    if (!this.token) {
      throw new Error("No token");
    }

    init = {
      ...init,
      headers: {
        ...(init ? init.headers : {}),
        Authorization: `Bearer ${this.token}`
      }
    };

    return fetch(input, init);
  }

  public saveToken(token: string) {
    localStorage.setItem("token", token);
    this.token = token;
    this.parseToken();
  }

  public checkToken(token: string): boolean {
    return this.token === token;
  }

  public clearToken() {
    localStorage.removeItem("token");
    this.token = null;
    this.parsedToken = undefined;
  }

  public getName() {
    return this.realName;
  }

  public updateName(name: string) {
    this.realName = name;
    if (this.nameChanged) {
      this.nameChanged(name);
    }
  }

  public isAdmin() {
    if (this.parsedToken) {
      return this.parsedToken.role === "Admin";
    }
    return false;
  }

  public hasToken(): boolean {
    return !!this.token;
  }

  public isInDate() {
    if (this.parsedToken) {
      return isFuture(this.parsedToken.exp * 1000);
    }

    return false;
  }

  private parseToken() {
    if (this.token) {
      try {
        this.parsedToken = jwtDecode(this.token);
      } catch {
        this.token = null;
        this.updateName("");
        localStorage.removeItem("token");
        this.parsedToken = undefined;
      }

      if (this.parsedToken) {
        this.updateName(this.parsedToken.given_name);
      }
    } else {
      this.parsedToken = undefined;
      this.updateName("");
    }
  }
}

const fetcher = new Fetcher();

export { fetcher, Fetcher };
