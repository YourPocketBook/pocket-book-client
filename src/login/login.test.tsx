import "@testing-library/jest-dom/extend-expect";
import { fireEvent, render, wait } from "@testing-library/react";
import fetchMock from "fetch-mock";
import React from "react";
import ReactDOM from "react-dom";
import { Login } from ".";
import "../_testSupport/fetch-helpers";
import { inDateNonAdminToken } from "../_testSupport/tokens";
import { fetcher } from "../fetcher";
import { history } from "../history";

const endpoint = "/api/login";

describe("Login Page", () => {
  beforeEach(() => {
    history.push = jest.fn();
  });

  afterEach(() => {
    fetcher.clearToken();
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Login />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders default correctly", () => {
    const { container } = render(<Login />);

    expect(container).toMatchSnapshot();
  });

  it("defaults to correct values", () => {
    const { getByLabelText } = render(<Login />);

    expect(getByLabelText("Email")).toHaveAttribute("value", "");
    expect(getByLabelText("Password")).toHaveAttribute("value", "");
  });

  it("declares valid values correctly", async () => {
    const testEmail = "test@sja.org.uk";
    const testPassword = "testPassword";

    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        h.body === JSON.stringify({ email: testEmail, password: testPassword }),
      {
        body: JSON.stringify([{ path: "doses", error: "test" }]),
        status: 400
      }
    );

    const { getByLabelText, getByText } = render(<Login />);

    fireEvent.change(getByLabelText("Email"), {
      target: { value: testEmail }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testPassword }
    });

    fireEvent.click(getByText("Login", { selector: "button" }));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByLabelText("Email")).toHaveClass("is-valid");
    expect(getByLabelText("Password")).toHaveClass("is-valid");
  });

  it("declares invalid values correctly", () => {
    const { getByLabelText, getByText } = render(<Login />);

    fireEvent.change(getByLabelText("Email"), {
      target: { value: "test-value" }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: "" }
    });

    fireEvent.click(getByText("Login", { selector: "button" }));

    expect(getByLabelText("Email")).toHaveClass("is-invalid");
    expect(getByLabelText("Password")).toHaveClass("is-invalid");
  });

  it("declares non-sja email invalid correctly", () => {
    const { getByLabelText, getByText } = render(<Login />);

    fireEvent.change(getByLabelText("Email"), {
      target: { value: "test@gmail.com" }
    });

    fireEvent.click(getByText("Login", { selector: "button" }));

    expect(getByLabelText("Email")).toHaveClass("is-invalid");
  });

  it("renders server error correctly", async () => {
    const testEmail = "test@sja.org.uk";
    const testPassword = "testPassword";

    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        h.body === JSON.stringify({ email: testEmail, password: testPassword }),
      { status: 500 }
    );

    const { getByText, getByLabelText } = render(<Login />);

    fireEvent.change(getByLabelText("Email"), {
      target: { value: testEmail }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testPassword }
    });
    fireEvent.click(getByText("Login", { selector: "button" }));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText(/The server has had an issue/)).toBeInTheDocument();
  });

  it("renders in progress correctly", () => {
    const testEmail = "test@sja.org.uk";
    const testPassword = "testPassword";

    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        h.body === JSON.stringify({ email: testEmail, password: testPassword }),
      // tslint:disable-next-line: no-empty
      new Promise(() => {})
    );
    const { getByText, getByLabelText } = render(<Login />);

    fireEvent.change(getByLabelText("Email"), {
      target: { value: testEmail }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testPassword }
    });
    fireEvent.click(getByText("Login", { selector: "button" }));

    expect(getByText("Login", { selector: "button" })).toBeDisabled();
  });

  it("renders not authorised correctly", async () => {
    const testEmail = "test@sja.org.uk";
    const testPassword = "testPassword";

    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        h.body === JSON.stringify({ email: testEmail, password: testPassword }),
      { status: 401 }
    );

    const { getByText, getByLabelText } = render(<Login />);

    fireEvent.change(getByLabelText("Email"), {
      target: { value: testEmail }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testPassword }
    });
    fireEvent.click(getByText("Login", { selector: "button" }));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(
      getByText(/That username and password was not recognised/)
    ).toBeInTheDocument();
  });

  it("renders not confirmed correctly", async () => {
    const testEmail = "test@sja.org.uk";
    const testPassword = "testPassword";

    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        h.body === JSON.stringify({ email: testEmail, password: testPassword }),
      {
        body: JSON.stringify([{ error: "emailNotConfirmed", path: "email" }]),
        status: 400
      }
    );

    const { getByText, getByLabelText } = render(<Login />);

    fireEvent.change(getByLabelText("Email"), {
      target: { value: testEmail }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testPassword }
    });
    fireEvent.click(getByText("Login", { selector: "button" }));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText(/Your email hasn't been confirmed/)).toBeInTheDocument();
  });

  it("toggles show password when button is clicked", () => {
    const { getByText, getByLabelText } = render(<Login />);

    fireEvent.click(getByText("Show Password"));
    expect(getByLabelText("Password")).toHaveAttribute("type", "text");

    fireEvent.click(getByText("Hide Password"));
    expect(getByLabelText("Password")).toHaveAttribute("type", "password");
  });

  it("redirects on successful form submit", async () => {
    const testEmail = "test@sja.org.uk";
    const testPassword = "testPassword";

    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        h.body === JSON.stringify({ email: testEmail, password: testPassword }),
      {
        body: JSON.stringify({ token: inDateNonAdminToken }),
        status: 200
      }
    );

    const { getByLabelText, getByText } = render(<Login />);

    fireEvent.change(getByLabelText("Email"), {
      target: { value: testEmail }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testPassword }
    });
    fireEvent.click(getByText("Login", { selector: "button" }));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(history.push).toBeCalledWith("/home");
    expect(fetcher.checkToken(inDateNonAdminToken)).toBeTruthy();
  });
});
