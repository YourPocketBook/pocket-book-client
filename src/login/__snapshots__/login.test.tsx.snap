// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page renders default correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
    </nav>
    <div
      class="centreContainer container"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-6 offset-lg-3"
        >
          <div
            class="centreBox"
          >
            <div
              class="container"
            >
              <form
                class=""
                id="login-form"
                novalidate=""
              >
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <h1
                      class="text-center"
                    >
                      Login
                    </h1>
                  </div>
                </div>
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="form-group"
                    >
                      <label
                        class=""
                        for="email"
                      >
                        Email
                      </label>
                      <input
                        autocomplete="email"
                        class="form-control"
                        id="email"
                        name="email"
                        placeholder="Email"
                        type="email"
                        value=""
                      />
                      <div
                        class="invalid-feedback"
                      >
                        You need to give your SJA email.
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="form-group"
                    >
                      <label
                        class=""
                        for="password"
                      >
                        Password
                      </label>
                      <input
                        autocomplete="current-password"
                        class="form-control"
                        id="password"
                        name="password"
                        placeholder="Password"
                        type="password"
                        value=""
                      />
                      <div
                        class="invalid-feedback"
                      >
                        You need to enter a password.
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="row"
                >
                  <div
                    class="text-right col-md-9 offset-md-3"
                  >
                    <button
                      class="text-muted btn btn-link"
                      id="showPassword"
                      type="button"
                    >
                      Show Password
                    </button>
                  </div>
                </div>
                <button
                  class="btn btn-dark-green btn-block"
                  type="submit"
                >
                  Login
                </button>
                <a
                  class="btn btn-light-green btn-block"
                  href="/forgot-password"
                  id="log-in-button"
                >
                  Reset Password
                </a>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright Â© 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
